docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,253 @@
+scenario we were not even able to connect more than 120 devices to two LNSs, while in a single LNS
+topology we could scale up to 900 devices without issues.
+
+## Out of scope
+
+- Deduplication strategies Mark and None: these strategies rely on multiple LNSs sending message.
+Potentially we could consider other work arounds for the IoT Hub limitation of a single connection
+per device but we find it acceptable for the Mark and None strategies to not be as scalable as the
+Drop strategy and will only document this limitation for potential users to be aware of.
+
+- LNS performs operations on behalf of a device/sensor and a concentrator/station. However since a
+concentrator can be connected to at most one LNS, there is no ping-pong happening with operations on
+stations.
+  
+## In-scope
+
+- The problem can be manifested whenever we do operations against Iot Hub on behalf of edge devices.
+  These can be:
+  - Twin reads
+  - Twin writes (updates/deletes)
+  - D2C messages
+  - C2D messages
+- Roaming leaf devices (that potentially become out-of-range from an LNS) are in scope.
+
+### Problematic IoT Hub operations on behalf of edge devices
+
+- Background tasks
+  - Periodically we refresh the LoRaDeviceCache, which results in device twin reads that could
+    switch the connection -> see [handling of background tasks section](#handling-of-background-tasks)
+- Message flows
+  - Join -> see [handling of Join requests section](#Handling-of-Join-requests)
+  - Data:
+    - if the device is not in LoRaDeviceCache, we fetch the device twin -> see [LoRaDevice not in cache section](#data-flow-loradevice-not-in-loradevicecache)
+    - if a frame counter reset happened, we update the twin immediately -> see [handling of resets section](#handling-of-device-resets)
+    - in the main data flow we send upstream, downstream and write the new twin -> see [main data flow section](#main-data-message-flow)
+    - Class C direct downstream messages sent from the portal could result in a connection switch if
+    the handling LNS changed in the meantime but we consider this as out of scope.
+
+Version, LNS discovery and CUPS update endpoints are not affected.
+
+## Possible solution
+
+The main idea is to give the current connection holder as indicated from the Function, the edge to
+continue processing messages for this device. That gateway, will have 0 impact on performance.
+
+### Handling of background tasks
+
+When we create a (singleton) instance of the NetworkServer.LoRaDeviceCache we start a background periodical
+check to ensure the device twins for the DevEuis we encountered are kept fresh.
+
+- If we were previously the losing LNS for the DevEui that needs refresh, we could adjust the
+  LastSeen property but not actually refresh the entry
+  - this way we won't try to read the twin that could result in a connection switch
+  - when the next data message comes and we check the cache
```

Comment:
I would be more specific here to indicate you are indicating the case of a primary gateway switch

Reply:
Was added explicitly that we are now the winning LNS.